mimic/model/identity.py

```diff
@@ -20,3 +26,205 @@ def from_store(cls, session_store, tenant_id):
+        """
+        return session_store.session_for_username_password(
+            self.username, self.password, self.tenant_id)
+
+    @classmethod
+    def from_json(cls, json_blob):
+        """
+        Given an authentication JSON blob, which should look like:
+
+            {
+                "auth": {
+                    "passwordCredentials": {
+                        "username": "user",
+                        "password": "pass"
+                    },
+                    "tenantId": "111111"
+                }
+            }
+
+        ``"tenantId"`` is interchanable with ``"tenantName"``, and is
+        optional.
+
+        :return: a class:`PasswordCredentials` object
+        """
+        tenant_id = (json_blob['auth'].get('tenantName', None) or
+                     json_blob['auth'].get('tenantId', None))
+        username = json_blob['auth']['passwordCredentials']['username']
+        password = json_blob['auth']['passwordCredentials']['password']
+        return cls(username=username, password=password, tenant_id=tenant_id)
+
+
+@implementer(ICredentials)
+@attributes
+class APIKeyCredentials(object):
+    """
+    An object representing username/api-key + optional tenant credentials
+    """
+    username = attr(validator=validators.instance_of(string_types))
+    api_key = attr(validator=validators.instance_of(string_types))
+    tenant_id = attr(default=None)
+
+    def get_session(self, session_store):
+        """
+        Get a session corresponding to the user and tenant from the given
+        session store.
+
+        :param session_store: a :class:`mimic.session.SessionStore`
+        """
+        return session_store.session_for_api_key(
+            self.username, self.api_key, self.tenant_id)
+
+    @classmethod
+    def from_json(cls, json_blob):
+        """
+        Given an authentication JSON blob, which should look like:
+
+            {
+                "auth": {
+                    "RAX-KSKEY:apiKeyCredentials": {
+                        "username": "user",
+                        "apiKey": "key"
+                    },
+                    "tenantId": "111111"
+                }
+            }
+
+        ``"tenantId"`` is interchanable with ``"tenantName"``, and is
+        optional.
+
+        :return: a class:`APIKeyCredentials` object
+        """
+        creds = json_blob['auth']
+        tenant_id = (creds.get('tenantName', None) or
+                     creds.get('tenantId', None))
+        username = creds['RAX-KSKEY:apiKeyCredentials']['username']
+        api_key = creds['RAX-KSKEY:apiKeyCredentials']['apiKey']
+        return cls(username=username, api_key=api_key, tenant_id=tenant_id)
+
+
+@implementer(ICredentials)
+@attributes
+class TokenCredentials(object):
+    """
+    An object representing token + optional tenant credentials
+    """
+    token = attr(validator=validators.instance_of(string_types))
+    tenant_id = attr(validator=validators.instance_of(string_types))
```

Comment:
It does not support nullability.  You have to provide either a `tenantId` or `tenantName`:  http://docs.rackspace.com/auth/api/v2.0/auth-client-devguide/content/POST_authenticate_token_v2.0_tokens_Token_Calls.html